breathecode/payments/actions.py

```diff
@@ -526,30 +533,49 @@ def get_amount(bag: Bag, currency: Currency, lang: str) -> tuple[float, float, f
     if not currency:
         currency, _ = Currency.objects.get_or_create(code="USD", name="United States dollar")
 
-    for service_item in bag.service_items.all():
-        if service_item.service.currency != currency:
-            bag.service_items.remove(service_item)
-            continue
-
-        price_per_month += service_item.service.price_per_unit * service_item.how_many
-        price_per_quarter += service_item.service.price_per_unit * service_item.how_many * 3
-        price_per_half += service_item.service.price_per_unit * service_item.how_many * 6
-        price_per_year += service_item.service.price_per_unit * service_item.how_many * 12
-
     for plan in bag.plans.all():
         if plan.currency != currency:
             bag.plans.remove(plan)
             continue
 
         must_it_be_charged = ask_to_add_plan_and_charge_it_in_the_bag(plan, user, lang)
 
-        # this prices is just used if it are generating a subscription
         if not bag.how_many_installments and (bag.chosen_period != "NO_SET" or must_it_be_charged):
             price_per_month += plan.price_per_month or 0
             price_per_quarter += plan.price_per_quarter or 0
             price_per_half += plan.price_per_half or 0
             price_per_year += plan.price_per_year or 0
 
+    plans = bag.plans.all()
+    add_ons: dict[int, AcademyService] = {}
+    for plan in plans:
+        for add_on in plan.add_ons.filter(currency=currency):
+            if add_on.service.id not in add_ons:
+                add_ons[add_on.service.id] = add_on
+
+    for service_item in bag.service_items.all():
+        if service_item.service.id in add_ons:
+            add_on = add_ons[service_item.service.id]
+
+            try:
+                add_on.validate_transaction(service_item.how_many, lang)
+            except Exception as e:
+                bag.service_items.filter().delete()
+                bag.plans.filter().delete()
+                raise e
+
+            if price_per_month != 0:
+                price_per_month += add_on.get_discounted_price(service_item.how_many) * 1
+
+            if price_per_quarter != 0:
+                price_per_quarter += add_on.get_discounted_price(service_item.how_many) * 3
+
+            if price_per_half != 0:
+                price_per_half += add_on.get_discounted_price(service_item.how_many) * 6
+
+            if price_per_year != 0:
+                price_per_year += add_on.get_discounted_price(service_item.how_many) * 12
```

Comment:
Don't multiply all the service's prices as we are only going to charge them once, regardless the subscription time